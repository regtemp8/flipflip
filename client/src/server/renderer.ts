import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
