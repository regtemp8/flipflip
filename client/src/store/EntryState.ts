export interface Identifiable {
  id: number
}

export interface EntryState<T extends Identifiable> {
  name: string
  nextID: number
  entries: Record<number, T>
}

export interface EntryUpdate<T> {
  id: number
  value: T
}

export function isEntry<T extends Identifiable>(
  state: EntryState<T>,
  id: number
) {
  return state.entries[id] != null
}

export function setNextID<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  if (state.nextID <= entry.id) {
    state.nextID = entry.id + 1
  }
}

export function getEntry<T extends Identifiable>(
  state: EntryState<T>,
  id: number
): T {
  return state.entries[id]
}

export function setEntry<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  state.entries[entry.id] = entry
  setNextID(state, entry)
}

export function createEntry<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  if (!isEntry(state, entry.id)) {
    state.entries[entry.id] = entry
    setNextID(state, entry)
  } else {
    throw new Error(
      `${state.name}: Create failed. Entry with ID ${entry.id} already exists.`
    )
  }
}

export function createEntries<T extends Identifiable>(
  state: EntryState<T>,
  entries: T[]
) {
  for (const entry of entries) {
    if (isEntry(state, entry.id)) {
      throw new Error(
        `${state.name}: Create failed. Entry with ID ${entry.id} already exists.`
      )
    }
  }

  entries.forEach((s) => (state.entries[s.id] = s))
  const maxID = Math.max(...entries.map((s) => s.id))
  setNextID(state, { id: maxID })
}

export function updateEntry<T extends Identifiable>(
  state: EntryState<T>,
  entry: T
) {
  if (isEntry(state, entry.id)) {
    state.entries[entry.id] = entry
  } else {
    throw new Error(
      `${state.name}: Update failed. Entry with ID ${entry.id} doesn't exist.`
    )
  }
}

export function deleteEntry<T extends Identifiable>(
  state: EntryState<T>,
  id: number
) {
  if (isEntry(state, id)) {
    delete state.entries[id]
  } else {
    throw new Error(
      `${state.name}: Delete failed. Entry with ID ${id} doesn't exist.`
    )
  }
}
