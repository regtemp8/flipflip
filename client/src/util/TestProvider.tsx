import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
