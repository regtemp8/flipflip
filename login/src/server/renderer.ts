import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
