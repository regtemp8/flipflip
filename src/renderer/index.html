<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
