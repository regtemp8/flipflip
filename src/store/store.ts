import { configureStore } from '@reduxjs/toolkit'
import appSlice from './app/slice'
import audioSlice from './audio/slice'
import constantsSlice from './constants/slice'
import playersSlice from './player/slice'
import captionScriptSlice from './captionScript/slice'
import clipSlice from './clip/slice'
import librarySourceSlice from './librarySource/slice'
import overlaySlice from './overlay/slice'
import playlistSlice from './playlist/slice'
import sceneSlice from './scene/slice'
import sceneGridSlice from './sceneGrid/slice'
import sceneGroupSlice from './sceneGroup/slice'
import tagSlice from './tag/slice'
import videoClipperSlice from './videoClipper/slice'
import scenePickerSlice from './scenePicker/slice'
import captionScriptorSlice from './captionScriptor/slice'
import sceneDetailSlice from './sceneDetail/slice'
import audioLibrarySlice from './audioLibrary/slice'
import sourceScraperSlice from './sourceScraper/slice'

const store = configureStore({
  reducer: {
    app: appSlice,
    audio: audioSlice,
    captionScript: captionScriptSlice,
    clip: clipSlice,
    constants: constantsSlice,
    librarySource: librarySourceSlice,
    overlay: overlaySlice,
    playlist: playlistSlice,
    scene: sceneSlice,
    sceneGrid: sceneGridSlice,
    sceneGroup: sceneGroupSlice,
    tag: tagSlice,
    // components
    players: playersSlice,
    videoClipper: videoClipperSlice,
    scenePicker: scenePickerSlice,
    captionScriptor: captionScriptorSlice,
    sceneDetail: sceneDetailSlice,
    audioLibrary: audioLibrarySlice,
    sourceScraper: sourceScraperSlice
  }
})

export default store
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
