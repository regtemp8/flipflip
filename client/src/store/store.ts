/// <reference path="../global.d.ts"/>
import { createReduxStore } from '../server/renderer'

const store = createReduxStore(
  window.flipflipAppStorage,
  window.flipflipConstants
)

export default store
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
