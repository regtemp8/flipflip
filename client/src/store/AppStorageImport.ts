import { initialApp } from "./app/data/App";
import { initialAudioState } from "./audio/slice";
import { initialCaptionScriptState } from "./captionScript/slice";
import { initialClipState } from "./clip/slice";
import { initialLibrarySourceState } from "./librarySource/slice";
import { initialOverlayState } from "./overlay/slice";
import { initialPlaylistState } from "./playlist/slice";
import { initialSceneState } from "./scene/slice";
import { initialSceneGridState } from "./sceneGrid/slice";
import { initialSceneGroupState } from "./sceneGroup/slice";
import { initialTagState } from "./tag/slice";
import { EntryState } from "./EntryState";
import type Tag from "./tag/Tag";
import type SceneGroup from "./sceneGroup/SceneGroup";
import type SceneGrid from "./sceneGrid/SceneGrid";
import type Scene from "./scene/Scene";
import type Playlist from "./playlist/Playlist";
import type Overlay from "./overlay/Overlay";
import type LibrarySource from "./librarySource/LibrarySource";
import type Clip from "./clip/Clip";
import type CaptionScript from "./captionScript/CaptionScript";
import type Audio from "./audio/Audio";
import type App from "./app/data/App";

export interface AppStorageImport {
  app: App;
  audio: EntryState<Audio>;
  captionScript: EntryState<CaptionScript>;
  clip: EntryState<Clip>;
  librarySource: EntryState<LibrarySource>;
  overlay: EntryState<Overlay>;
  playlist: EntryState<Playlist>;
  scene: EntryState<Scene>;
  sceneGrid: EntryState<SceneGrid>;
  sceneGroup: EntryState<SceneGroup>;
  tag: EntryState<Tag>;
}

export const initialAppStorageImport: AppStorageImport = {
  app: initialApp,
  audio: initialAudioState,
  captionScript: initialCaptionScriptState,
  clip: initialClipState,
  librarySource: initialLibrarySourceState,
  overlay: initialOverlayState,
  playlist: initialPlaylistState,
  scene: initialSceneState,
  sceneGrid: initialSceneGridState,
  sceneGroup: initialSceneGroupState,
  tag: initialTagState,
}