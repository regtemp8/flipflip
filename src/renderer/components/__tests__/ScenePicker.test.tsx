import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import ScenePicker from "../ScenePicker";
import TestProvider from "../../../../test/util/TestProvider";
import store from "../../../store/store";

jest.mock('../../animations/Jiggle', () => 'Jiggle');
jest.mock('../../animations/VSpin', () => 'VSpin');
jest.mock('../SceneSearch', () => 'SceneSearch');

// mocking this so that test doesn't throw error
jest.mock('react-sortablejs', () => 'Sortable');

// TODO create functional tests instead of snapshots
describe("ScenePicker", () => {
  it("should match snapshot", () => {
    const component = renderer.create(
      <TestProvider store={store}>
        <ScenePicker />
      </TestProvider>
    );

    let tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
  });
});