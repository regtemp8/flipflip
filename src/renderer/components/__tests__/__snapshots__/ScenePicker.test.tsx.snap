// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScenePicker should match snapshot 1`] = `
<div
  className="ScenePicker-root-1"
  onClick={[Function]}
>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionAbsolute ScenePicker-appBar-2 css-iu2t21-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <button
        aria-label="Toggle Drawer"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <VSpin>
        <div
          className="ScenePicker-logo-4"
        />
      </VSpin>
      <h1
        className="MuiTypography-root MuiTypography-h4 MuiTypography-noWrap ScenePicker-title-6 css-1r3v6n7-MuiTypography-root"
      >
        FlipFlip
      </h1>
      <span
        className="MuiTypography-root MuiTypography-caption MuiTypography-noWrap ScenePicker-version-7 css-44w6i2-MuiTypography-root"
      >
        v
      </span>
      <div
        className="ScenePicker-fill-44"
      />
      <SceneSearch
        displaySources={[]}
        filters={[]}
        onUpdateFilters={[Function]}
        placeholder="Search ..."
      />
    </div>
  </header>
  <div
    className="MuiDrawer-root MuiDrawer-docked css-7ik032-MuiDrawer-docked"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper ScenePicker-drawerPaper-9 ScenePicker-drawerPaperClose-10 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-12i7wg6-MuiPaper-root-MuiDrawer-paper"
    >
      <div
        className="ScenePicker-drawerToolbar-11"
      >
        <button
          aria-label="Toggle Drawer"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <VSpin>
          <div
            className="ScenePicker-logo-4 ScenePicker-drawerLogo-5"
          />
        </VSpin>
        <h1
          className="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-2l1qg7-MuiTypography-root"
        >
          FlipFlip
        </h1>
      </div>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div>
        <div
          className="MuiTabs-root MuiTabs-vertical ScenePicker-tabs-38 css-1g4ac1o-MuiTabs-root"
        >
          <div
            className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            onScroll={[Function]}
            style={
              {
                "marginRight": -0,
                "overflow": "hidden",
              }
            }
          >
            <div
              aria-label="scene picker tabs"
              aria-orientation="vertical"
              className="MuiTabs-flexContainer MuiTabs-flexContainerVertical css-lfwcke-MuiTabs-flexContainer"
              onKeyDown={[Function]}
              role="tablist"
            >
              <button
                aria-controls="vertical-tabpanel-0"
                aria-selected={true}
                className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected ScenePicker-tab-39 ScenePicker-sceneTab-41 ScenePicker-tabClose-40 css-1i8cmzc-MuiButtonBase-root-MuiTab-root"
                disabled={false}
                id="vertical-tab-0"
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="tab"
                tabIndex={0}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="MovieIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m18 4 2 4h-3l-2-4h-2l2 4h-3l-2-4H8l2 4H7L5 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4h-4z"
                  />
                </svg>
                
                <span
                  className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="vertical-tabpanel-1"
                aria-selected={false}
                className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary ScenePicker-tab-39 ScenePicker-generatorTab-42 ScenePicker-tabClose-40 css-1i8cmzc-MuiButtonBase-root-MuiTab-root"
                disabled={false}
                id="vertical-tab-1"
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="MovieFilterIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m18 4 2 3h-3l-2-3h-2l2 3h-3l-2-3H8l2 3H7L5 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4h-4zm-6.75 11.25L10 18l-1.25-2.75L6 14l2.75-1.25L10 10l1.25 2.75L14 14l-2.75 1.25zm5.69-3.31L16 14l-.94-2.06L13 11l2.06-.94L16 8l.94 2.06L19 11l-2.06.94z"
                  />
                </svg>
                
                <span
                  className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="vertical-tabpanel-2"
                aria-selected={false}
                className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary ScenePicker-tab-39 ScenePicker-gridTab-43 ScenePicker-tabClose-40 css-1i8cmzc-MuiButtonBase-root-MuiTab-root"
                disabled={false}
                id="vertical-tab-2"
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="tab"
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="GridOnIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zM8 20H4v-4h4v4zm0-6H4v-4h4v4zm0-6H4V4h4v4zm6 12h-4v-4h4v4zm0-6h-4v-4h4v4zm0-6h-4V4h4v4zm6 12h-4v-4h4v4zm0-6h-4v-4h4v4zm0-6h-4V4h4v4z"
                  />
                </svg>
                
                <span
                  className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              className="MuiTabs-indicator css-t3lg8-MuiTabs-indicator"
              style={
                {
                  "height": 0,
                  "top": 0,
                }
              }
            />
          </div>
        </div>
      </div>
      <div
        className="ScenePicker-fill-44"
      />
      <div
        className="ScenePicker-drawerBottom-12 ScenePicker-drawerBottomClose-13"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 css-16epbcf-MuiTypography-root"
        >
          Questions? Suggestions?
          <br />
          Visit us on 
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17biv4x-MuiTypography-root-MuiLink-root"
            href="#"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
          >
            GitHub
          </a>
           or 
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17biv4x-MuiTypography-root-MuiLink-root"
            href="#"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
          >
            Reddit
          </a>
        </p>
      </div>
    </div>
  </div>
  <main
    className="ScenePicker-content-17"
  >
    <div
      className="ScenePicker-appBarSpacer-16"
    />
    <div
      className="MuiContainer-root ScenePicker-container-18 css-ur2jdm-MuiContainer-root"
    >
      <div
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        <Sortable
          className="ScenePicker-sceneList-19"
          onChange={[Function]}
          options={
            {
              "animation": 150,
              "easing": "cubic-bezier(1, 0, 0, 1)",
              "group": {
                "name": "sg.scene",
                "pull": true,
                "put": true,
              },
            }
          }
        />
        <Sortable
          onChange={[Function]}
          options={
            {
              "animation": 150,
              "easing": "cubic-bezier(1, 0, 0, 1)",
              "group": {
                "name": "group",
                "pull": false,
                "put": false,
              },
              "handle": ".group-handle",
            }
          }
        />
      </div>
    </div>
  </main>
  <button
    aria-label="Random Scene"
    aria-labelledby={null}
    className="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-default ScenePicker-randomButton-25 ScenePicker-extraWindowRandomButton-26 css-12m9wit-MuiButtonBase-root-MuiFab-root"
    data-mui-internal-clone-element={true}
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium ScenePicker-icon-36 css-i4bv87-MuiSvgIcon-root"
      data-testid="CasinoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM7.5 18c-.83 0-1.5-.67-1.5-1.5S6.67 15 7.5 15s1.5.67 1.5 1.5S8.33 18 7.5 18zm0-9C6.67 9 6 8.33 6 7.5S6.67 6 7.5 6 9 6.67 9 7.5 8.33 9 7.5 9zm4.5 4.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm4.5 4.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm0-9c-.83 0-1.5-.67-1.5-1.5S15.67 6 16.5 6s1.5.67 1.5 1.5S17.33 9 16.5 9z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
